File: RTS/Assets/Standard Assets/FogOfWar/FogOfWar.js

var newMaterial : Material;
var targetAlpha = 0.75;
var alphaAnimationSpeed = 0.5;

var gridSize = 3.0;
var xLength = 10;
var yLength = 10;


static var RAYCAST_HEIGHT = 50.0;

private var verticesToReveal = new Array();
private var mesh : Mesh;
private var colors : Color[];

private var resettingFog = false;
private var revealedVerts = new Array(); // the verts that need to be darkened again
private var initialRevealVerts = new Array(); // the verts that don't get darkened again since there are units on them
private var revealers = new Array();

function AddVert(pos : Vector3, x : float, z : float) : Vector3
{
	var hitInfo : RaycastHit;
	var layerMask = 1 << 10;
	Physics.Raycast(Vector3(x+pos.x,RAYCAST_HEIGHT,z+pos.z), -Vector3.up, hitInfo, Mathf.Infinity, layerMask);
	//Debug.Log((x+pos.x) + ", " + hitInfo.point.y + ", " + (z+pos.z));
	return Vector3(x,hitInfo.point.y,z);
}

function Awake()
{
	gameObject.AddComponent("MeshFilter");
	gameObject.AddComponent("MeshRenderer");
	mesh = GetComponent(MeshFilter).mesh;

	mesh.Clear();

	// vertices ---------------------------------------------------------------------------------
	var verts = new Array();

	var pos = transform.position;
	//var locZ = transform.position.z;

	for (var y = 0; y < yLength; ++y)
	{
		for (var x = 0; x < xLength; ++x)
		{
			verts.Push( AddVert(pos,(x*gridSize),(y*gridSize)) );
		}
	}
	mesh.vertices = verts.ToBuiltin(Vector3);

	// uv's ----------------------------------------------------------------------------------------
	var uvs = new  Vector2[verts.length];

	for (var i=0;i<uvs.Length;i++) {
		uvs[i] = Vector2(verts[i].x, verts[i].z);
	}
	mesh.uv = uvs;

	// triangles ---------------------------------------------------------------------------------
	var tris = new Array();
	//var trilen = (mesh.vertices.length-2)/2;

	for (y = 0; y < yLength-1; ++y)
	{
		for (x = 0; x < xLength-1; ++x)
		{
			var vertHere = x+(y*xLength);
			var vertAbove = x+((y+1)*xLength);
			tris.Push(vertHere); 
			tris.Push(vertAbove);
			tris.Push(vertAbove+1);

			tris.Push(vertHere);
			tris.Push(vertAbove+1);
			tris.Push(vertHere+1);
			//Debug.Log(x + "," + y + ": " +
			//	vertHere + "," + vertAbove + "," + (vertAbove+1) + " " +
			//	vertHere + "," + (vertAbove+1) + "," + (vertHere+1));
		}
	}
	mesh.triangles = tris.ToBuiltin(int);
	mesh.RecalculateNormals();
	mesh.RecalculateBounds();

	// vertex colors ------------------------------------------------------------------------------
	var vertices = mesh.vertices;
	colors = new Color[vertices.Length];

	for (i=0;i<vertices.Length;i++)
	{
		colors[i] = Color(0,0,0,targetAlpha);
	}
	mesh.colors = colors;


	renderer.castShadows = false;
	renderer.receiveShadows = false;
	renderer.material = newMaterial;
	
	gameObject.AddComponent("MeshCollider");
	var meshCollider : MeshCollider = GetComponent(MeshCollider);
	meshCollider.sharedMesh = mesh;
}

// same as Array.Push but doesn't
// push it if it already exists in that Array
function UniquePush(arr : Array, elementToAdd)
{
	var putIn = true;
	for (var i = 0; i < arr.length; ++i)
	{
		if (arr[i] == elementToAdd)
		{
			// already in, no need to put
			putIn = false;
			break;
		}
	}

	if (putIn)
	{
		arr.Push(elementToAdd);
	}
}


function AddVertToReveal(vertIdx : int, quickReveal : boolean)
{
	if (quickReveal)
	{
		if (colors[vertIdx].a >= targetAlpha)
		{
			colors[vertIdx].a = 0;
			mesh.colors = colors;
		}
		
		UniquePush(initialRevealVerts, vertIdx);
		//Debug.Log("initialRevealVerts now: " + initialRevealVerts.length);
	}
	else
	{
		if (colors[vertIdx].a <= 0.0)
		{
			return;
		}

		UniquePush(verticesToReveal, vertIdx);
	}
}

function IsVertRevealed(vertIdx : int) : boolean
{
	return (colors[vertIdx].a < targetAlpha);
}

function Update()
{
	//Debug.Log(verticesToReveal.length);

	if (verticesToReveal.length > 0)
	{
		for (var i = 0; i < verticesToReveal.length; ++i)
		{
			var vertIdxToReveal = verticesToReveal[i];

			colors[vertIdxToReveal].a -= alphaAnimationSpeed * Time.deltaTime;
			if (colors[vertIdxToReveal].a <= 0.0)
			{
				verticesToReveal.RemoveAt(i);
				if (!resettingFog)
				{
					UniquePush(revealedVerts, vertIdxToReveal);
				}
				--i;
				if (i <= -1)
				{
					i = 0;
				}
			}
		}
	}

	if (resettingFog)
	{
		if (revealedVerts.length == 0)
		{
			resettingFog = false;
		}
		else
		{
			//Debug.Log("darkening");
			for (i = 0; i < revealedVerts.length; ++i)
			{
				var vertIdxToDarken = revealedVerts[i];

				colors[vertIdxToDarken].a += alphaAnimationSpeed * Time.deltaTime;
				if (colors[vertIdxToDarken].a >= targetAlpha)
				{
					revealedVerts.RemoveAt(i);
					--i;
					if (i <= -1)
					{
						i = 0;
					}
				}
			}
		}
	}

	if ((verticesToReveal.length > 0) || resettingFog)
	{
		mesh.colors = colors;
	}
}



function ResetFog()
{
	// add contents of initialRevealVerts to revealedVerts
	for (idx in initialRevealVerts)
	{
		UniquePush(revealedVerts, idx);
	}

	// clear initialRevealVerts
	initialRevealVerts.Clear();
	//Debug.Log("initialRevealVerts now: " + initialRevealVerts.length);

	// tell all revealers to call their InitialReveal to repopulate initialRevealVerts
	//Debug.Log("revealers: " + revealers.length);
	for (r in revealers)
	{
		r.InitialReveal();
	}

	//Debug.Log("initialRevealVerts after all revealers to call their InitialReveal: " + initialRevealVerts.length);

	// remove elements of initialRevealVerts from revealedVerts
	// revealedVerts - initialRevealVerts
	for (initial in initialRevealVerts)
	{
		for (var i = 0; i < revealedVerts.length; ++i)
		{
			if (revealedVerts[i] == initial)
			{
				revealedVerts.RemoveAt(i);
				--i;
				if (i <= -1)
				{
					i = 0;
				}
			}
		}
	}


	// redarken verts found in revealedVerts
	resettingFog = true;

	//for (vertIdx in revealedVerts)
	//{
	//	colors[vertIdx].a = targetAlpha;
	//}
	//revealedVerts.Clear();
	//mesh.colors = colors;
}



function AddRevealer(r : FogOfWarRevealer)
{
	UniquePush(revealers, r);
}

File: trunk/RTS/Assets/Scripts/Misc/FogOfWar.cs
using UnityEngine;
using System.Collections.Generic;

[AddComponentMenu("Misc RTS/Fog Of War")]
public class FogOfWar : MonoBehaviour {
	
	public Material newMaterial;
	public float undiscoveredAlpha = 1f;
	public float fogAlpha = 0.5f;
	public float alphaAnimationSpeed = 0.8f;
	
	public float gridSize = 5.0f;
	public int xLength = 10;
	public int yLength = 10;
	
	public static float RAYCAST_HEIGHT = 50.0f;
	
	protected Mesh mesh;
	protected List<Color> colors;
	
	// Provides a public shortcut to the mesh's triangles array, since calling mesh.triangles
	// actually creates a copy of that huge array for every access
	public int[] meshTriangles;
	
	// Holds all verts that became visible to a unit
	protected List<int> verticesToReveal = new List<int>();
	// Holds all verts that are fully revealed
	protected List<int> revealedVerts = new List<int>(); // the verts that need to be darkened again
	
	// Builds the fog mesh and colors list based on the settings provided
	void Awake() {
		gameObject.AddComponent<MeshFilter>();
		gameObject.AddComponent<MeshRenderer>();
		mesh = GetComponent<MeshFilter>().mesh;
	
		mesh.Clear();
	
		// vertices ---------------------------------------------------------------------------------
		List<Vector3> verts = new List<Vector3>();
	
		Vector3 pos = transform.position;
		//var locZ = transform.position.z;
	
		for (int y = 0; y < yLength; ++y) {
			for (int x = 0; x < xLength; ++x) {
				verts.Add( AddVert(pos,(x*gridSize),(y*gridSize)) );
			}
		}
		mesh.vertices = verts.ToArray();
	
		// uv's ----------------------------------------------------------------------------------------
		List<Vector2> uvs = new List<Vector2>(verts.Count);
	
		for (int i=0; i<verts.Count; i++) {
			uvs.Add(new Vector2(verts[i].x, verts[i].z));
		}
		mesh.uv = uvs.ToArray();
	
		// triangles ---------------------------------------------------------------------------------
		List<int> tris = new List<int>();
	
		for (int y = 0; y < yLength-1; ++y) {
			for (int x = 0; x < xLength-1; ++x) {
				int vertHere = x+(y*xLength);
				int vertAbove = x+((y+1)*xLength);
				tris.Add(vertHere); 
				tris.Add(vertAbove);
				tris.Add(vertAbove+1);
	
				tris.Add(vertHere);
				tris.Add(vertAbove+1);
				tris.Add(vertHere+1);
			}
		}
		mesh.triangles = tris.ToArray();
		meshTriangles = tris.ToArray();
		mesh.RecalculateNormals();
		mesh.RecalculateBounds();
	
		// vertex colors ------------------------------------------------------------------------------
		colors = new List<Color>(mesh.vertices.Length);
	
		for (int i=0; i<mesh.vertices.Length; i++) {
			colors.Add(new Color(0, 0, 0, undiscoveredAlpha));
		}
		mesh.colors = colors.ToArray();
	
	
		renderer.castShadows = false;
		renderer.receiveShadows = false;
		renderer.material = newMaterial;
		
		gameObject.AddComponent("MeshCollider");
		MeshCollider meshCollider = GetComponent<MeshCollider>();
		meshCollider.sharedMesh = mesh;
	}
	
	
	Vector3 AddVert(Vector3 pos, float x, float z) {
		RaycastHit hitInfo;
		int layerMask = 1 << LayerMask.NameToLayer("FogOfWar");
		Physics.Raycast(new Vector3(x+pos.x, RAYCAST_HEIGHT, z+pos.z), -Vector3.up, out hitInfo, Mathf.Infinity, layerMask);
		return new Vector3(x, hitInfo.point.y, z);
	}
	
	// Adds a vert that should be revealed because it is visible to some unit
	public void AddVertToReveal(int vertIdx) {
		if (colors[vertIdx].a > 0.0f && !verticesToReveal.Contains(vertIdx)) {
			verticesToReveal.Add(vertIdx);
		}
	}
	
	public bool IsVertRevealed(int vertIdx) {
		return (colors[vertIdx].a < fogAlpha);
	}
	
	public bool IsVertDiscovered(int vertIdx) {
		return (colors[vertIdx].a < undiscoveredAlpha);
	}
	
	void Update() {
		// Darken all revealed verts
		for (int i = 0; i < revealedVerts.Count; ++i) {
			int vertIdxToDarken = revealedVerts[i];
			
			/*
			 * to optimize, we could keep a dictionary of verts with a list of non-moving aivisions that can see them
			 * then ignore verts in that dictionary to darken. that way non-moving aivisions wouldn't have to recalculate their vision every frame
			 */
			
			// Except don't darken a vert if it is supposed to be revealed right now
			if(!verticesToReveal.Contains(vertIdxToDarken)) {
				Color color = colors[vertIdxToDarken];
				color.a += alphaAnimationSpeed * Time.deltaTime;
				colors[vertIdxToDarken] = color;
				if(color.a >= fogAlpha) {
					// Vert is totally darkened, so remove it from revealedVerts
					revealedVerts.RemoveAt(i);
					--i;
					if(i <= -1) {
						i = 0;
					}
				}
			}
		}
		
		// And lighten all visible ones
		for (int j = 0; j < verticesToReveal.Count; ++j) {
			int vertIdxToReveal = verticesToReveal[j];
			
			// Except don't lighten a vert if it is already totally revealed
			if(!revealedVerts.Contains(vertIdxToReveal)) {
				Color color = colors[vertIdxToReveal];
				color.a -= alphaAnimationSpeed * Time.deltaTime;
				colors[vertIdxToReveal] = color;
				if(color.a <= 0.0f) {
					// Vert is totally transparent, so remove it from verticesToReveal and add it to revealedVerts
					if(!revealedVerts.Contains(vertIdxToReveal)) {
						revealedVerts.Add(vertIdxToReveal);
					}
					verticesToReveal.RemoveAt(j);
					--j;
					if(j <= -1) {
						j = 0;
					}
				}
			} else {
				verticesToReveal.RemoveAt(j);
			}
		}
		
		mesh.colors = colors.ToArray();
	}

}